DataFormats/Common/interface/StdArray.h

```diff
@@ -0,0 +1,270 @@
+  // See https://github.com/root-project/root/issues/12007 for a discussion on the issue.
+  //
+  // This class reimplements the full std::array<T,N> interface, using a regular
+  // Reflex dictionary for the ROOT serialisation.
+  // To be more GPU-friendly, all methods are constexpr, and out-of-bound data access
+  // aborts instead of throwing an exception.
+  //
+  // Note: dictonaries for edm::StdArray<T,N> where T is a standard C/C++ type
+  // should be declared in DataFormats/Common/src/classed_def.xml.
+
+  namespace detail {
+    template <typename T, std::size_t N>
+    class StdArrayTrait {
+    public:
+      using array_type = T[N];
+    };
+
+    template <typename T>
+    class StdArrayTrait<T, 0> {
+    public:
+      struct array_type {};
+    };
+  }  // namespace detail
+
+  template <typename T, std::size_t N>
+  class StdArray {
+  public:
+    // Member types
+
+    using value_type = T;
+    using size_type = std::size_t;
+    using difference_type = std::ptrdiff_t;
+    using reference = value_type&;
+    using const_reference = value_type const&;
+    using pointer = value_type*;
+    using const_pointer = const value_type*;
+    using iterator = pointer;
+    using const_iterator = const_pointer;
+    using reverse_iterator = std::reverse_iterator<iterator>;
+    using const_reverse_iterator = std::reverse_iterator<const_iterator>;
+
+    // Constructors
+
+    // default constructor
+    constexpr StdArray() = default;
+
+    // copy constructor
+    constexpr StdArray(StdArray<T, N> const& init) = default;
+
+    // move constructor
+    constexpr StdArray(StdArray<T, N>&& init) = default;
+
+    // copy assignment operator
+    constexpr StdArray& operator=(StdArray<T, N> const& init) = default;
+
+    // move assignment operator
+    constexpr StdArray& operator=(StdArray<T, N>&& init) = default;
+
+    // Interoperability with std::array
+
+    // copy assignment from an std::array
+    constexpr StdArray& operator=(std::array<T, N> const& init) {
+      for (size_type i = 0; i < N; ++i) {
+        data_[i] = init[i];
+      }
+      return *this;
+    }
+
+    // move assignment from an std::array
+    constexpr StdArray& operator=(std::array<T, N>&& init) {
+      for (size_type i = 0; i < N; ++i) {
+        data_[i] = std::move(init[i]);
+      }
+      return *this;
+    }
+
+    // cast operator to an std::array
+    constexpr operator std::array<T, N>() const {
+      std::array<T, N> copy;
+      for (size_type i = 0; i < N; ++i) {
+        copy[i] = data_[i];
+      }
+      return copy;
+    }
+
+    // Element access
+
+    // Returns a reference to the element at specified location pos, with bounds checking.
+    // If pos is not within the range of the container, the program aborts.
+    constexpr reference at(size_type pos) {
+      if (pos >= N)
+        abort();
+      return data_[pos];
+    }
+    constexpr const_reference at(size_type pos) const {
+      if (pos >= N)
```

Comment:
Do you want to add
```suggestion
      if constexpr (pos >= N)
```
here too?